internal/util-complete/src/main/scala/sbt/internal/util/LineReader.scala

```diff
@@ -8,24 +8,98 @@
 package sbt.internal.util
 
 import java.io._
+import java.util.{ List => JList }
 
 import jline.console.ConsoleReader
 import jline.console.history.{ FileHistory, MemoryHistory }
+import org.jline.reader.{
+  Candidate,
+  Completer,
+  EndOfFileException,
+  LineReader => JLineReader,
+  LineReaderBuilder,
+  ParsedLine,
+  UserInterruptException,
+}
 import sbt.internal.util.complete.Parser
 
 import scala.annotation.tailrec
 import scala.concurrent.duration._
+import java.nio.channels.ClosedByInterruptException
 
-trait LineReader {
+trait LineReader extends AutoCloseable {
   def readLine(prompt: String, mask: Option[Char] = None): Option[String]
+  override def close(): Unit = {}
 }
 
 object LineReader {
   val HandleCONT =
     !java.lang.Boolean.getBoolean("sbt.disable.cont") && Signals.supported(Signals.CONT)
   val MaxHistorySize = 500
 
+  private def completer(parser: Parser[_]): Completer = new Completer {
+    def complete(lr: JLineReader, pl: ParsedLine, candidates: JList[Candidate]): Unit = {
+      Parser.completions(parser, pl.line(), 10).get.foreach { c =>
+        /*
+         * For commands like `~` that delegate parsing to another parser, the `~` may be
+         * excluded from the completion result. For example,
+         * ~testOnly <TAB>
+         * might return results like
+         * 'testOnly ;'
+         * 'testOnly com.foo.FooSpec'
+         * ...
+         * If we use the raw display, JLine will reject the completions because they are
+         * missing the leading `~`. To workaround this, we append to the result to the
+         * line provided the line does not end with " ". This fixes the missing `~` in
+         * the prefix problem. We also need to split the line on space and take the
+         * last token and append to that otherwise the completion will double print
+         * the prefix, so that `testOnly com<Tab>` might expand to something like:
+         * `testOnly testOnly\ com.foo.FooSpec` instead of `testOnly com.foo.FooSpec`.
+         */
+        if (c.append.nonEmpty) {
+          if (!pl.line().endsWith(" ")) {
+            candidates.add(new Candidate(pl.line().split(" ").last + c.append))
+          } else {
+            candidates.add(new Candidate(c.append))
+          }
+        }
+      }
+    }
+  }
   def createReader(
+      historyPath: Option[File],
+      parser: Parser[_],
+      terminal: Terminal,
+      prompt: Prompt = Prompt.Running,
+  ): LineReader = {
+    val term = JLine3(terminal)
+    // We may want to consider insourcing LineReader.java from jline. We don't otherwise
+    // directly need jline3 for sbt.
```

Comment:
I'm a bit confused by this statement. LineReader.java (https://github.com/jline/jline3/blob/3b0d5d52fd2c31fb92ad8aaa2f11a22aceab890a/reader/src/main/java/org/jline/reader/LineReader.java) is an empty interface. Don't we need JLine's implementation for the reader?

Reply:
Yeah. I meant the implementation class https://github.com/jline/jline3/blob/3b0d5d52fd2c31fb92ad8aaa2f11a22aceab890a/reader/src/main/java/org/jline/reader/impl/LineReaderImpl.java. The idea would just be to make a partial fork of jline3 that only provides the minimal subset of functionality from `LineReaderImpl` that we need to make line reading and tab completion work but without using all of other jline 3 stuff that we neither want or need. I'm not sure how excited I am about actually doing that though.

Reply:
What's the diff for `show sbtProj/Compile/externalDependencyClasspath` pre and post this PR?

Reply:
```patch
+ Attributed($MAVEN/org/jline/jline/3.15.0/jline-3.15.0.jar)
```